### PR TITLE
IEN-912 | End Of Journey - Complete

### PR DESCRIPTION
**Pull request by jerry-ey**
## Changes
1. add a new lambda with cron job at 9:00 am UTC (1 or 2 am PST) every day
2. add 2 migrations: First adding 2 new status. Second adding status_id in table `active_flag`
3. lambda will trigger `endOfJourneyService` to check, see the following detail description:
(Additionally, all queries are wrapped in a transaction to ensure they execute correctly.)
4. Fronted showing "archived status" if "showing hidden applicant" toggle is `on`
---

#### 1. Get Data
This update adds a query to retrieve applicant status audits with specific criteria. The query selects applicants who:
- Have an accepted job offer status.
- Do not have a corresponding entry in the `ien_applicants_active_flag` table (applicants without an active flag).
- Have a non-null `effective_date`.

The query is built using the following logic:

1. **Select Fields**:
   - `applicant_id` from the `IENApplicantStatusAudit` table.
   - `effective_date`, formatted as `YYYY-MM-DD`, selecting the maximum date.
   - `status` from the `status` relation.
   - `ha_pcn_id` from the `job` relation.

2. **Joins**:
   - Left joins the `status` and `job` tables to retrieve related status and job information.
   - Left joins the `ien_applicants_active_flag` table to exclude applicants already in this table by checking for `NULL` values on `active_flag.applicant_id`.

4. **Filters**:
   - Filters for applicants whose status is `JOB_OFFER_ACCEPTED`.
   - Excludes any applicants already in the `ien_applicants_active_flag` table.
   - Ensures `effective_date` is not null.

5. **Grouping and Aggregation**:
   - Groups by `applicant_id`, with additional grouping on `status.id` and `job.id`.
   - Uses a `HAVING` clause to ensure the `effective_date` is exactly one year before yesterday.

This query provides a filtered view of applicants with an accepted job offer, ensuring no duplicate or active entries in `ien_applicants_active_flag`.

#### 2. Set Data

The `setCompletedLists` function processes a list of applicants and performs the following operations:

1. **Audit Record Creation**:
   - For each applicant in the `list`, an entry is added to the `IENApplicantStatusAudit` table with:
     - `applicant_id`: Set from the applicant data in the list.
     - `start_date`: Set to today's date in `YYYY-MM-DD` format based on the `America/Los_Angeles` timezone.
     - `status`: Assigned the milestone status `END_OF_JOURNEY_COMPLETE`.
     - `notes`: Includes a timestamp and a message indicating that the status was updated by the Lambda CRON job.

2. **Update or Insert in `ien_applicants_active_flag`**:
   - The function attempts to retrieve the `status_id` for `END_OF_JOURNEY_COMPLETE`. If not found, it logs an error and throws an exception.
   - For each applicant:
     - It tries to update the `ien_applicants_active_flag` table to set `is_active` to `false` for the applicant's `ha_pcn_id`.
     - If no rows were updated (indicating no existing record), it inserts a new entry in `ien_applicants_active_flag` with `is_active` set to `false` and the retrieved `status_id`.

3. **Applicant Record Update**:
   - Finally, for each applicant, the function updates the `ien_applicants` table, setting the `status` to `END_OF_JOURNEY_COMPLETE` and updating the `updated_date` to the current date.

This function ensures that applicants who have completed their journey are appropriately marked in the system, with entries in both the audit and active flag tables reflecting the `END_OF_JOURNEY_COMPLETE` milestone.

---

## Screenshots
- Lambda
CleanShot 2024-10-28 at 13 20 04

- Frontend showing inactive status
CleanShot 2024-10-28 at 12 54 46

## TESTING
- start testing on DEV, adding a JOB accepted on Oct 28th, 2024
CleanShot 2024-10-28 at 13 23 37

## TODO
~~- make sure the query not effected by running multiple times~~
- add testing file

### CODE REVIEW

**Comment by jerry-ey** [comment]
@dbayly-freshworks Please review the preliminary code.

### For Testing the code: 
- I have deployed to DEV and added a new milestone JOB OFFER ACCEPTED with `Target start date: 2023-10-28`.
So I will check tomorrow see if the cron job is working fine. Expecting it will add new milestone: `End of Jounery - Journey Complete` to this applicant and archive it.
CleanShot 2024-10-28 at 13 31 16
- I will also add api e2e testings later.

File: apps/api/src/migration/1729889576502-AddEndOfJourneyApplicantStatus.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class AddEndOfJourneyApplicantStatus1729889576502 implements MigrationInterface {
+  milestones = [
```
Comment (dbayly-freshworks) [inline apps/api/src/migration/1729889576502-AddEndOfJourneyApplicantStatus.ts:4]: We'll need to update our documentation to include these IDs in our list of milestones.

**Review by dbayly-freshworks — APPROVED** [review]
Overall looks good - A few comments

File: apps/api/src/migration/1729889576502-AddEndOfJourneyApplicantStatus.ts (hunk 1 above)
Comment (jerry-ey) [reply apps/api/src/migration/1729889576502-AddEndOfJourneyApplicantStatus.ts:4]: Added to Wiki:
https://github.com/bcgov/internationally-educated-nurses/wiki/Documentation

**Pull request merged**